### PR TITLE
Using fields for custom data types instead of mixins

### PR DESCRIPTION
**Pull request by tswicegood**
After a talk with @jacobian the other day in Lawrence, I came back with the idea of using [custom fields](http://docs.djangoproject.com/en/1.3/howto/custom-model-fields/) throughout our standard models to give people the ability to standardize on a set of features and functionality.  These commits are some prototyping in that direction to get some feedback.

_Note_: These are not complete as an implementation of a `VideoField`, just an idea to get the ball rolling and get some feedback on what people think of the field vs. mixin setup.  For a comparison, see [arm_content.authorship.models.AuthorshipMixin](https://github.com/niran/armstrong.core.arm_content/blob/master/armstrong/core/arm_content/authorship/models.py).
### Things this is missing
- Does not store the type/backend.  Need to determine if we want to add another field dynamically to the model to handle that, or actually require that it get explicitly added to the model
- No ability to search based on type
- Needs at least a `VimeoBackend`, possibly others still
- Needs to support multiple backends
- Needs to be able to handle backends that don't like data
- Should have an `embed` or `html` property for grabbing the embed code (backend should know how to do this and/or where to fetch the oembed data from)

### CODE REVIEW

**Comment by tswicegood** [comment]
As a quick, more apples to apples comparison, to change `AuthorshipMixin` to `AuthorsField`, you would probably write it something like:

```
authors = models.AuthorsField(override='my_byline_override', extension='byline_extra')
```

`override` and `extension` could look for the default fields, but would just treat them as not existing if they weren't on the model.  You would interact with it slightly differently too.  Like the `ExternalVideo` object in this pull request, you'd end up with a `ModelAuthors` object or some such that has an `html` property that renders the HTML version of the author's list, while casting `model.authors` to a string would generate the plain text version.

**Comment by tswicegood** [comment]
Hmm... Not sure why I didn't catch this already, but `VideoField` should actually be `ExternalVideoField` or `EmbeddedVideoField` to leave room for real videos with real files behind them.  Would be really cool if a real `VideoField` could also be hooked up to several upload backends so Vimeo, YouTube, Facebook, etc., all got copies of the real video that was uploaded, but I'll tag that as a pony. :-)

**Comment by tinio** [comment]
Choosing between the two names,  I prefer EmbeddedVideoField.   I think it conveys more the purpose of the field.

**Comment by tinio** [comment]
In regards to

> - Does not store the type/backend. Need to determine if we want to add another field dynamically to the model to handle that, or actually require that it get explicitly added to the model

I'm leaning towards adding the field dynamically.  Thinking, we can pull this up as the "Provider Name" from the oembed data.

**Comment by tinio** [comment]
In terms of the various backends we should support:

> - Needs at least a VimeoBackend, possibly others still
> - Needs to support multiple backends

I'm thinking an [EmbedlyBackend](http://embed.ly/) would be good in order to support a large number of providers.

**Comment by tinio** [comment]
> Should have an embed or html property for grabbing the embed code (backend should know how to do this and/or where to fetch the oembed data from)

I'd slightly prefer naming the property 'embed' as oppose to 'html'.  There are more services now (such as Storify) where the code to include is javascript.  It's technically wrapped within an html <script> tag, but there's a bit of a disconnect in that users only expect to enter a blob of html.  Fyi... at The Bay Citizen we currently have our property named 'html' and earlier I had to clarify this on a few occasions.  We now use the verbose_name 'embed' and there hasn't been confusion that this means the embed code.

**Comment by tswicegood** [comment]
I'm ok with `embed`.

I'm afraid by including the list of random things that weren't done, though, I miscommunicated what I was trying to do with this pull request.  I included the list to make sure everyone didn't think this was a final implementation.

More than anything I'm curious for input on this style of custom fields versus using mixins.  I'm finishing up the `AuthorsField` which is a more complex example and gives us a chance for a side-by-side comparison with the mixin approach that @niran has implemented.

**Comment by tinio** [comment]
My mistake.

You were clear on soliciting feedback about 'custom fields' vs 'mixins' but figured I'd offer comments on the other items as well.  I look forward to comparing against the `AuthorsField`.

Apologies and also Thanks

**Comment by tswicegood** [comment]
No worries -- just didn't want to get bogged down in implementation details.
 I've pushed some sketches of the `AuthorsField`.  As you can see, it is
basically a specialized `OneToOneField`, but does give a declarative way to
add that "field" and its functionality to any model without going down the
inheritance path.

**Comment by tinio** [comment]
+1 On use of custom fields throughout

Admittedly at first I wasn't sure how custom fields can work for all the various fields we would need.  However, having taken a look at the examples given here, more document reading and lots of time with the code under django.db.models.fields I do think the use of custom fields is the way to go vs mixins.  Even in our Content model now the proliferation of helper methods while still manageable has always been annoying and felt unnatural (too much of a hack).  While a bit more tricky to set up, I think the benefits of a cleaner namespace with custom fields outweigh the cons.  Also, I'd anticipate most of the fields we'd add to be variants of the ones already provided by Django, so that simplifies and helps with implementation a lot.

File: armstrong/core/arm_content/models.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+def user_to_name(user):
+    return user.get_full_name()
+
+
+class Authors(models.Model):
+    users = models.ManyToManyField(User)
+    override = models.CharField(max_length=200)
+    extra = models.CharField(max_length=200)
+
+    def __unicode__(self, formatter=user_to_name):
```
Comment (niran) [inline armstrong/core/arm_content/models.py:23]: This doesn't implement the same byline extension behavior as described in the comments of the `AuthorshipMixin` code. `" and "` should never be automatically added if an extension is present.

It also doesn't return `unicode`.

**Comment by niran** [comment]
The `AuthorsField` stuff looks reasonable, though I think the intermediate `Authors` model might cause admin headaches.

**Comment by tswicegood** [comment]
I don't think it will.  I intend on providing an `AuthorsInline` that we can pull in for handling authors directly on any particular post.

File: armstrong/core/arm_content/models.py (hunk 1 above)
Comment (tswicegood) [inline armstrong/core/arm_content/models.py:23]: FWIW: it doesn't need to return unicode.  Python is smart enough to treat the string appropriately.  See 4686ace515b569817d8c7699183a107cd42e97e2.  Those tests pass without the `u""` convention, but I did add it for additional clarity

Comment (niran) [inline armstrong/core/arm_content/models.py:23]: Even if it casts the end result to `unicode`, any authors with non-ASCII characters in their names would have caused a `UnicodeDecodeError`.

Comment (niran) [inline armstrong/core/arm_content/models.py:23]: Also, the spacing logic from `AuthorshipMixin` is missing as well. See https://github.com/niran/armstrong.core.arm_content/commit/65a968409af5bd1f829866a41be9a38ee1491cbc#L1R32.

Comment (tswicegood) [inline armstrong/core/arm_content/models.py:23]: On the spacing, yup... missed that one.

On the unicode, the `UnicodeDecodeError`: if that's the case why does 4686ace515b569817d8c7699183a107cd42e97e2 not fail?  That's the explicit case of an author with unicode in their name.  It didn't fail on 2.7.x, or am I missing something?

**Comment by tswicegood** [comment]
Ok, I've incorporated the feedback specific to these.  Unless something else is posted between now and the end of our 1 meeting, I'm going to close this (without a merge) and start focusing on the implementation.  I'll open pull requests for both `EmbeddedVideoField` and `AuthorsField` so everyone can weigh in on the implementation details of each.

**Comment by tswicegood** [comment]
Closing...

File: armstrong/core/arm_content/models.py (hunk 1 above)
Comment (niran) [inline armstrong/core/arm_content/models.py:23]: You're right. `join` will make everything unicode if any of the strings being joined are unicode.

Comment (tswicegood) [inline armstrong/core/arm_content/models.py:23]: I wish I could star comments.  "You're right."  Almost brought a tear to my eye... :-D

**Pull request closed without merging**